data/trie/patriciaMerkleTrie.go

```diff
@@ -657,3 +681,17 @@ func (tr *patriciaMerkleTrie) GetOldRoot() []byte {
 
 	return tr.oldRoot
 }
+
+// Close calls all the canceFuncs
+func (tr *patriciaMerkleTrie) Close() error {
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	for _, cancelFunc := range tr.cancelFuncs {
```

Comment:
Please do not use a map of cancel functions as this is sub-optimal
Instead use a single unbuffered chan of type struct{} that will be closed on the Close function.
Something like:
```GO
func (tr *patriciaMerkleTrie) Close() error {
     close(tr.chanClose)
     return nil
}
```

then the `getAllLeavesOnChannel` will get the tr.chanClose as parameter. All calls that might take some time will be done as:
```GO
for ...... {
    select{
    case <- chanClose:
          return
    default:
    }

    //operation
}

```
This is how I've done it in other places in elrond-go

Reply:
Nice pattern! Refactored.